docs/okeps/okep-5224-connecting-udns/okep-5224-connecting-udns.md

```diff
@@ -0,0 +1,1292 @@
+
+1) It is responsible to allocate a /31 range for each node-network pair:
+  * It get's the `connectSubnet` value from the CNC and allocates
+    a `/31` point to point link subnet for each node-network pair
+  * These values are annotated on the `ClusterNetworkConnect`
+    CRD for it to be then consumed on the `ovnkube-controller` side.
+    Sample:
+    ```yaml
+    k8s.ovn.org/node-network-subnets: {
+      L3Network1Node1: 192.168.0.0/31,
+      L3Network1Node2: 192.168.0.2/31,
+      L3Network1Node3: 192.168.0.4/31,
+      L3Network2Node1: 192.168.0.6/31,
+      L3Network2Node2: 192.168.0.8/31,
+      L3Network2Node3: 192.168.0.10/31,
+      L2Network1: 192.168.0.12/31,
+      L2Network2: 192.168.0.14/31,
+      ...
+    }
+    ```
+    We will need N×M subnets for N networks and M nodes for Layer3
+    and N×1 subnets for Layer2 for N layer2 networks.
+2) It is responsible for allocating the global tunnel key for each
+  `connect-router`:
+  * It will use the same instance of global `TunnelKeysAllocator`
+    in cluster manager that already allocates the tunnel keys for
+    existing Layer2 transit routers. This allocator already
+    preserves the first 4096 IDs which are used as tunnel-keys
+    for Layer2 transit switches. It will then allocate the new ones
+    for the connect router dynamically. This ensures we don't reuse
+    already allocated networkIDs. For each connect API that will
+    create a transit router, we will need 1 tunnel key per router.
+  * This value is annotated on the `ClusterNetworkConnect`
+    CRD for it to be then consumed on the `ovnkube-controller` side.
+    Sample:
+    ```yaml
+    k8s.ovn.org/connect-router-tunnel-key: '50000'
+    ```
+4) Status conditions will be updated based on feedback of setup
+ from the `ovnkube-controllers` in different zones.
+
+This controller also needs to watch for UDNs, CUDNs and Nodes to
+accommodate adds, updates and deletes.
+
+**OVN-Kubernetes Controller**
+
+A new `network-connect-controller` will be created that watches the
+`ClusterNetworkConnect` objects and does the following:
+
+1) Based on the network selectors, creates the
+   necessary network topology changes mentioned
+   in the above section:
+   * Creates the `connect-router`
+   * Creates the ports connecting each Layer3 `ovn_cluster_router` and
+     each Layer2 `transit-router` to the `connect-router`
+   * Adds the necessary logical router policies on each network router
+   * Adds the necessary logical router static routes on the `connect-router`
+2) Reads the `k8s.ovn.org/node-network-subnets`
+   annotation set by the cluster manager and allocates
+   an IP from /31 to each point to point router port link
+   and uses that IP to create the ports
+3) Reads the `k8s.ovn.org/connect-router-tunnel-key` and uses
+   the tunnel key to create the `connect-router`
+4) Within each `connect-router`, we would need N*M tunnel keys? TBD clarify
+   this with dceara.
```

Comment:
@tssurya Yes, on the "connect-router" we need:

one logical router port connected to each L2 UDN transit router that's part of the "cluster-network-connect" (so not one per node) => N tunnel keys if only L2 networks are interconnected
one logical router port connected to each L3 UDN node cluster router for each L3 UDN that's part of the "cluster-network-connect) => N x M tunnel keys if only L3 networks are interconnected.

A generic formula would be:

number_of_tunnel_keys (logical router ports) = N_L2 + N_L3 x M


where:

N_L2: number of L2 UDNs part of the cluster-network-connect
N_L3: number of L3 UDNs part of the cluster-network-connect
M: number of nodes in the cluster